go/logictest/runner.go

```diff
@@ -246,11 +246,40 @@ func runTestFile(harness Harness, file string) {
 	}
 }
 
+type R struct {
+	schema string
+	results []string
+	cont bool
+	err error
+}
+
 // Executes a single record and returns whether execution of records should continue
 func executeRecord(harness Harness, record *parser.Record) (schema string, results []string, cont bool, err error) {
-	startTime = time.Now()
 	currRecord = record
 
+	tc := make(chan *R, 1)
+	go func() {
+		schema, results, cont, err := execute(harness, record)
+		tc <- &R{
+			schema: schema,
+			results: results,
+			cont: cont,
+			err: err,
+		}
+	}()
+
+	select {
+	case res := <-tc:
+		return res.schema, res.results, res.cont, res.err
+	case <-time.After(1 * time.Hour):
```

Comment:
This is too long a timeout for a single record. There are thousands of records for select5.test, and most of them will time out. You really want a timeout per test file instead. An hour seems about right.

Reply:
After a test file hits the timeout, you probably should log timeout messages for every remaining test record in the file.

Reply:
So this is the right idea, but it's not sufficient. They way this is written, if the record ever completes, it will write to stdout as before, either duplicating test results for this record or even stomping on another record's results.

You need to use a context object that has a built-in timeout, and in execute() it needs to check the context for cancellation before logging anything.